featurebyte/migration/service/feature.py

```diff
@@ -0,0 +1,137 @@
+
+    This class is used to migrate the feature records to add relationships_info field.
+    """
+
+    # skip audit migration for this migration
+    skip_audit_migration = True
+
+    def __init__(
+        self,
+        persistent: Persistent,
+        feature_service: FeatureService,
+        entity_service: EntityService,
+        relationship_info_service: RelationshipInfoService,
+        entity_relationship_extractor_service: EntityRelationshipExtractorService,
+        derive_primary_entity_helper: DerivePrimaryEntityHelper,
+    ):
+        super().__init__(persistent)
+        self.feature_service = feature_service
+        self.entity_service = entity_service
+        self.relationship_info_service = relationship_info_service
+        self.entity_relationship_extractor_service = entity_relationship_extractor_service
+        self.derive_primary_entity_helper = derive_primary_entity_helper
+
+    @property
+    def delegate_service(self) -> BaseDocumentServiceT:
+        return self.feature_service  # type: ignore[return-value]
+
+    async def batch_preprocess_document(self, documents: List[Document]) -> List[Document]:
+        """
+        Preprocess the documents before migration
+
+        Parameters
+        ----------
+        documents: List[Document]
+            List of documents to be migrated
+
+        Returns
+        -------
+        List[Document]
+        """
+        all_entity_ids = set()
+        for document in documents:
+            all_entity_ids.update(document["entity_ids"])
+
+        # get all entity first to reduce the number of queries
+        entity_id_to_entity: Dict[ObjectId, EntityModel] = {
+            entity.id: entity
+            async for entity in self.entity_service.list_documents_iterator(
+                query_filter={"_id": {"$in": list(all_entity_ids)}}
+            )
+        }
+        extractor = self.entity_relationship_extractor_service
+        for document in documents:
+            primary_entity_ids = await self.derive_primary_entity_helper.derive_primary_entity_ids(
+                entity_ids=document["entity_ids"], entity_id_to_entity=entity_id_to_entity
+            )
+            relationships_info = await extractor.extract_relationship_from_primary_entity(
+                entity_ids=document["entity_ids"],
+                primary_entity_ids=primary_entity_ids,
+            )
+            document["primary_entity_ids"] = sorted(primary_entity_ids)
+            document["relationships_info"] = [
+                relationship.dict(by_alias=True) for relationship in relationships_info
+            ]
+        return documents
+
+    def migrate_document_record(self, record: Dict[str, Any]) -> Dict[str, Any]:
+        try:
+            output = super().migrate_document_record(record)
+            return output
+        except Exception as exc:  # pylint: disable=broad-except
+            # for some bad records, we just log the error and return the original record
```

Comment:
curious - do we know what the common causes for failure are?

Reply:
For the older records where `aggregation_ids` is missing (
https://github.com/featurebyte/featurebyte/blob/2aef0b53d0a51e23645dd275c8ff500743bc971d/featurebyte/models/feature.py#L452), the root validator attempts to derive the attribute. It fails to generate the attribute when deriving operation structure of the join node.